src/ordered-index.hpp

```diff
@@ -0,0 +1,183 @@
+    explicit ordered_index_t(std::size_t initial_block_size = 1024 * 1024)
+    : m_block_size(initial_block_size)
+    {}
+
+    /**
+     * This is the value returned from the getter functions if the id is not
+     * in the database.
+     */
+    static constexpr std::size_t not_found_value() noexcept
+    {
+        return std::numeric_limits<std::size_t>::max();
+    }
+
+    /**
+     * How many entries will fit into the currently allocated memory. This
+     * is accurate for normal operations, but if there are huge gaps between
+     * consecutive ids (> 2^32), less entries than this will fit.
+     */
+    std::size_t capacity() const noexcept;
+
+    /// The number of entries in the index.
+    std::size_t size() const noexcept { return m_size; }
+
+    /**
+     * Add an entry to the index.
+     *
+     * \param id The key of the index.
+     * \param offset The value of the index.
+     *
+     * \pre Id and offset must be larger than any previously added id and
+     *      offset, respectively.
+     */
+    void add(osmid_t id, std::size_t offset);
+
+    /**
+     * Get the offset for the specified id.
+     *
+     * If the id is not in the index \code not_found_value() \endcode is
+     * returned.
+     *
+     * \param id The id to look for.
+     */
+    std::size_t get(osmid_t id) const noexcept
+    {
+        auto const p = get_internal(id);
+        if (p.first != id) {
+            return not_found_value();
+        }
+        return p.second;
+    }
+
+    /**
+     * Get the offset for the specified id or, if the id is not in the index,
+     * the next smaller id available in the index.
+     *
+     * If the id is not in the index and no smaller id is in the index,
+     * \code not_found_value() \endcode is returned.
+     *
+     * \param id The id to look for.
+     */
+    std::size_t get_block(osmid_t id) const noexcept
+    {
+        return get_internal(id).second;
+    }
+
+    /**
+     * The approximate amount of bytes currently allocated by this index.
+     */
+    std::size_t used_memory() const noexcept
+    {
+        return m_ranges.capacity() * sizeof(range_entry) +
+               capacity() * sizeof(second_level_index_entry);
+    }
+
+    /**
+     * Clear all memory used by this index. The index can NOT be reused after
+     * that.
+     */
+    void clear()
+    {
+        m_ranges.clear();
+        m_ranges.shrink_to_fit();
+        m_size = 0;
+    }
+
+private:
+    struct second_level_index_entry
+    {
+        uint32_t id;
+        uint32_t offset;
+    };
+
+    struct range_entry
+    {
+        std::vector<second_level_index_entry> index;
+        osmid_t from;
+        osmid_t to = 0;
+        std::size_t offset_from;
+
+        range_entry(osmid_t id, std::size_t offset, std::size_t block_size)
+        : from(id), offset_from(offset)
+        {
+            index.reserve(block_size);
+        }
+
+        bool full() const noexcept { return index.size() == index.capacity(); }
```

Comment:
Is this guaranteed to work, i.e. is vector allowed to pre-emptively enlarge its capacity before it is full?

Reply:
https://en.cppreference.com/w/cpp/container/vector says iterators are only invalidated on, for example `push_back()` or `resize()` **if the vector changed capacity**. Documenting it this way makes only sense if we can know whether a vector changed capacity. So I think we are good here.